src/compute/src/render/join/delta_join.rs

```diff
@@ -315,14 +324,16 @@ fn build_halfjoin<G, Tr, CF>(
     closure: JoinClosure,
 ) -> (
     Collection<G, (Row, G::Timestamp), Diff>,
-    Collection<G, DataflowError, Diff>,
+    Vec<Collection<G, DataflowError, Diff>>,
 )
 where
     G: Scope,
     G::Timestamp: crate::render::RenderTimestamp,
     Tr: TraceReader<Time = G::Timestamp, Key = Row, Val = Row, R = Diff> + Clone + 'static,
     CF: Fn(&G::Timestamp, &G::Timestamp) -> bool + 'static,
 {
+    let mut errors = Vec::new();
+    let prev_key_could_error = prev_key.iter().any(|e| e.could_error());
     let (updates, errs) = updates.map_fallible("DeltaJoinKeyPreparation", {
```

Comment:
If we know that evaluating `prev_key` can't error, then we could use `map` instead of `map_fallible`